
# Exception handling
- Sever/application fails is Intolerable!
    - A resilient API will send back `4xx status code` when it receives a bad request, 
        - bad requests are `checked error`, they should be expected!
    - use `try/catch in controller`, or `field validation` to the data

- The `500 status code` should be reserved only for actual server/application failures: 
    - unexpected infrastracture scenarios,
    - high resource usage (CPU / Physical Memory)
    - database-related failure
- deal with Exceptions
  - `Try-Catch`: resolve in-place 
  - `ControllerAdvice` a central place to define all custom Exception Handler
## Try-Catch Exceptions
- handle common checked error inside the `Controllers`
    - id/resource not exist

```java
// Service method that risk of NoContactException
private int findIndexById(String id) throws NoContactException {
    return IntStream.range(0, contactRepository.getContacts().size())
        .filter(index -> contactRepository.getContacts().get(index).getId().equals(id))
        .findFirst()
        .orElseThrow(() -> new NoContactException());
}
// any Service method that use the findIndexById are at risk
Contact getContactById(String id) throws NoContactException;
void updateContact(String id, Contact contact) throws NoContactException;
void deleteContact(String id) throws NoContactException;

// the controller that uses these service should resolve the NoContactException
@GetMapping("/contact/{id}")
public ResponseEntity<Contact> getContact(@PathVariable String id) {
    try {
        Contact contact = contactService.getContactById(id);
        return new ResponseEntity<>(contact, HttpStatus.OK);
    } catch (NoContactException e) {
        return new ResponseEntity<>(null, null, HttpStatus.NO_CONTENT);
    }
}
// same for updateContact and deleteContact...
@PutMapping("/contact/{id}")
public ResponseEntity<Contact> updateContact(@PathVariable String id, @RequestBody Contact contact) {}
@DeleteMapping("/contact/{id}")
public ResponseEntity<HttpStatus> deleteContact(@PathVariable String id) {}
```

## ControllerAdvice 
- `@ControllerAdvice`: `class` annotation that allows you to `define global, custom exception handlers`.
    - `without the need to adding try-catch into multiple controllers!!!!`
    
- `@ExceptionHandler(Exception.class)`: `method` annotation that `defines an exception handler to handle a specific Exception`.

- `@ExceptionHandler` handles `RuntimeException` (unchecked). you should register your exception as RuntimeException
    - `RuntimeException` 不需要使用throw/throws声明和抛出; 由程序逻辑错误引起，可以由程序员test/debug进行修正。 

```java
// define the custom RuntimeException
public class ContactNotFoundException extends RuntimeException { 
    public ContactNotFoundException(String id) { /
        super("The id '" + id + "' does not exist in our records"); 
    }
}
```

- define global exception handler
    - Create a method called handleContactNotFoundException, and mark it as an `@ExceptionHandler.`
    - exception handler pick up exceptions of type `ContactNotFoundException as param`.
    - exception handler's `return type will be ResponseEntity<Object>`. It is common practice because Object means anything can be passed.
    - return a ResponseEntity that contains a status 404.
   
```java

@ControllerAdvice
public class ApplicationExceptionHandler {
    @ExceptionHandler(ContactNotFoundException.class)
    public ResponseEntity<Object> handleContactNotFoundException(ContactNotFoundException e) {
        HashMap<String, String> json = new HashMap<>();  
        // you can construct JSON on-fly using map. Alternatively use `JSONObject` 
        json.put("msg", e.getMessage());
        json.put("timestamp", LocalDateTime.now().toString());
        // can also use ErrorResponse
        ErrorResponse error = new ErrorResponse(Arrays.asList(ex.getLocalizedMessage())); 
        return new ResponseEntity<>(json, HttpStatus.NOT_FOUND);  // the client receive the hashMap as json
    }
}
```



- one handler can be used to handle multiple exceptions
```java
    @ExceptionHandler({ CourseNotFoundException.class, GradeNotFoundException.class, StudentNotFoundException.class })
    public ResponseEntity<Object> handleResourceNotFoundException(RuntimeException ex) {
        return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    }
```